streamingccl: tweak node lag replanning policy

msbutler opened the pull request:
This patch changes the policy for when pcr replans based on a lagging node:

- previously, lag node replanning would not occur if the replicated time had
  not advanced during the flow. This patch relaxes this requirement. If one
node is lagging behind all other nodes, and it is preventing the replicated
time from advancing, we really should replan!

- previously, lag node replanning would occur if the lagging node was more than
  x minutes behind the second most lagging node. But, if 2 nodes are lagging
behind all other nodes, we would not replan, even though we should. Instead,
this patch will trigger replanning if the lagging node's frontier is more than
x minutes behind the mean frontier of all participating nodes.

Overall, this patch relaxes the requirements for lag replanning, which will
become much less costly once #125044 lands, as it reduces the amount of wasted
work on each dist sql flow restart.

Informs #124915

Relase note: none

msbutler committed:
streamingccl: reset retry counter if persisted spans have updated

Previously, the job level retry counter would reset if the replicated time had
advanced. This patch relaxes this requirement a bit by reseting the retry
counter if the persisted spans have updated. This patch is motivated by a
subsequent patch that will enable more frequent distsql replanning when a node
lags behind.

Epic: none

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124926)
<!-- Reviewable:end -->

msbutler committed:
streamingccl: tweak node lag replanning policy

This patch changes the policy for when pcr replans based on a lagging node:

- previously, lag node replanning would not occur if the replicated time had
  not advanced during the flow. This patch relaxes this requirement. If one
node is lagging behind all other nodes, and it is preventing the replicated
time from advancing, we really should replan!

- previously, lag node replanning would occur if the lagging node was more than
  x minutes behind the second most lagging node. But, if 2 nodes are lagging
behind all other nodes, we would not replan, even though we should. Instead,
this patch will trigger replanning if the lagging node's frontier is more than
x minutes behind the mean frontier of all participating nodes.

Overall, this patch relaxes the requirements for lag replanning, which will
become much less costly once #125044 lands, as it reduces the amount of wasted
work on each dist sql flow restart.

Informs #124915

Relase note: none

msbutler commented:
unrelated flake

dt commented on pkg/ccl/streamingccl/streamingest/node_lag_detector.go:
```diff
@@ -15,32 +15,37 @@ import (
-	log.VEventf(ctx, 2, "computed min lag diff: %d lagging node, difference %.2f", laggingNode, minLagDifference.Minutes())
-	if maxAllowableLag < minLagDifference {
-		return errors.Wrapf(ErrNodeLagging, "node %d is %.2f minutes behind the next node. Try replanning", laggingNode, minLagDifference.Minutes())
+	laggingNode, maxLagDifference := computeMeanLagDifference(ctx, executionDetails)
```
max or mean?

dt commented on pkg/jobs/jobspb/jobs.go:
```diff
@@ -44,3 +44,22 @@ func (fes RestoreFrontierEntries) Equal(fes2 RestoreFrontierEntries) bool {
+// TODO(msbutler): use generics and combine with above.
+type ResolvedSpanEntries []ResolvedSpan
+
+func (fes ResolvedSpanEntries) Equal(fes2 ResolvedSpanEntries) bool {
```
"fes" ?

msbutler replied:
lol, i think i blindly copied from the method defined above it. Fixed.

msbutler committed:
streamingccl: deflake TestTenantStreamingJobRetryReset

Epic: none

Release note: none

msbutler commented:
TFTR!

bors r=dt
